configuration/postgres/postgres.go

```diff
@@ -0,0 +1,401 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4/pgxpool"
+	_ "github.com/jackc/pgx/v4/stdlib"
+
+	"github.com/dapr/components-contrib/configuration"
+	"github.com/dapr/kit/logger"
+)
+
+type ConfigurationStore struct {
+	metadata             metadata
+	client               *pgxpool.Pool
+	logger               logger.Logger
+	subscribeStopChanMap sync.Map
+	ActiveSubscriptions  map[string]*subscription
+}
+
+type subscription struct {
+	uuid    string
+	trigger string
+	keys    []string
+}
+
+const (
+	configtablekey               = "table"
+	connMaxIdleTimeKey           = "connMaxIdleTime"
+	connectionStringKey          = "connectionString"
+	ErrorMissingTableName        = "missing postgreSQL configuration table name"
+	InfoStartInit                = "initializing postgreSQL configuration store"
+	ErrorMissingConnectionString = "missing postgreSQL connection string"
+	ErrorAlreadyInitialized      = "PostgreSQL configuration store already initialized"
+	ErrorMissinMaxTimeout        = "missing PostgreSQL maxTimeout setting in configuration"
+	QueryTableExists             = "SELECT EXISTS (SELECT FROM pg_tables where tablename = $1)"
+	maxIdentifierLength          = 64 // https://www.postgresql.org/docs/current/limits.html
+	ErrorTooLongFieldLength      = "field name is too long"
+)
+
+func NewPostgresConfigurationStore(logger logger.Logger) configuration.Store {
+	logger.Debug("Instantiating PostgreSQL configuration store")
+	return &ConfigurationStore{
+		logger: logger,
+	}
+}
+
+func (p *ConfigurationStore) Init(metadata configuration.Metadata) error {
+	p.logger.Debug(InfoStartInit)
+
+	if p.client != nil {
+		return fmt.Errorf(ErrorAlreadyInitialized)
+	}
+	if m, err := parseMetadata(metadata); err != nil {
+		p.logger.Error(err)
+		return err
+	} else {
+		p.metadata = m
+	}
+	p.ActiveSubscriptions = make(map[string]*subscription)
+	ctx, cancel := context.WithTimeout(context.Background(), p.metadata.maxIdleTime)
+	defer cancel()
+	client, err := Connect(ctx, p.metadata.connectionString, p.metadata.maxIdleTime)
+	if err != nil {
+		return err
+	}
+	p.client = client
+	pingErr := p.client.Ping(ctx)
+	if pingErr != nil {
+		return pingErr
+	}
+	// check if table exists
+	exists := false
+	err = p.client.QueryRow(ctx, QueryTableExists, p.metadata.configTable).Scan(&exists)
```

Comment:
Two issues with this:

1. This doesn't check the value of `exists`
2. This should probably return a more descriptive error if the table doesn't exist. In particular, if the component can't automatically create the table, it should also link to the Dapr docs with instructions to create the table

Reply:
Scan raises an exception "ErrNoRows" if it does not find any rows with the error message as "no rows in result set". The documentation in Daprdocs includes table creation PR - https://github.com/dapr/docs/issues/2779

Reply:
Correct. Please detect `ErrNoRows` and show a more user-friendly error message, with a link to the Dapr docs

Reply:
Done